xline-client/src/clients/auth.rs

```diff
@@ -41,4 +55,697 @@ impl AuthClient {
+        }
+        self.handle_req_slow_path(request.inner).await
+    }
+
+    /// Gets role.
+    ///
+    /// # Errors
+    ///
+    /// If request fails to send
+    #[inline]
+    pub async fn role_get(&self, request: AuthRoleGetRequest) -> Result<AuthRoleGetResponse> {
+        self.handle_req_fast_path(request.inner).await
+    }
+
+    /// Lists role.
+    ///
+    /// # Errors
+    ///
+    /// If request fails to send
+    #[inline]
+    pub async fn role_list(&self) -> Result<AuthRoleListResponse> {
+        self.handle_req_fast_path(xlineapi::AuthRoleListRequest {})
+            .await
+    }
+
+    /// Deletes role.
+    ///
+    /// # Errors
+    ///
+    /// If request fails to send
+    #[inline]
+    pub async fn role_delete(
+        &self,
+        request: AuthRoleDeleteRequest,
+    ) -> Result<AuthRoleDeleteResponse> {
+        self.handle_req_slow_path(request.inner).await
+    }
+
+    /// Grants role permission.
+    ///
+    /// # Errors
+    ///
+    /// If request fails to send
+    #[inline]
+    pub async fn role_grant_permission(
+        &self,
+        request: AuthRoleGrantPermissionRequest,
+    ) -> Result<AuthRoleGrantPermissionResponse> {
+        if request.inner.perm.is_none() {
+            return Err(ClientError::InvalidArgs(String::from(
+                "Permission not given",
+            )));
+        }
+        self.handle_req_slow_path(request.inner).await
+    }
+
+    /// Revokes role permission.
+    ///
+    /// # Errors
+    ///
+    /// If request fails to send
+    #[inline]
+    pub async fn role_revoke_permission(
+        &self,
+        request: AuthRoleRevokePermissionRequest,
+    ) -> Result<AuthRoleRevokePermissionResponse> {
+        self.handle_req_slow_path(request.inner).await
+    }
+
+    /// Send request using fast path
+    async fn handle_req_fast_path<Req: Into<RequestWrapper>, Res: From<ResponseWrapper>>(
+        &self,
+        request: Req,
+    ) -> Result<Res> {
+        let propose_id = self.generate_propose_id();
+        let request = RequestWithToken::new_with_token(request.into(), self.token.clone());
+        let cmd = Command::new(vec![], request, propose_id);
+        let cmd_res = self.curp_client.propose(cmd).await?;
+        Ok(cmd_res.decode().into())
+    }
+
+    /// Send request using slow path
+    async fn handle_req_slow_path<Req: Into<RequestWrapper>, Res: From<ResponseWrapper>>(
+        &self,
+        request: Req,
+    ) -> Result<Res> {
+        let propose_id = self.generate_propose_id();
+        let request = RequestWithToken::new_with_token(request.into(), self.token.clone());
+        let cmd = Command::new(vec![], request, propose_id);
+        let (cmd_res, sync_res) = self.curp_client.propose_indexed(cmd).await?;
+        let mut res_wrapper = cmd_res.decode();
+        res_wrapper.update_revision(sync_res.revision());
+        Ok(res_wrapper.into())
+    }
```

Comment:
These two methods are very similar, I suggest merging them to reduce duplication

Reply:
done